### PR TITLE
Add metrics for azure service operations (route and loadbalancer).

### PR DESCRIPTION
**Pull request by nilo19**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Add metrics for azure service operations (route and loadbalancer).

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add metrics for azure service operations (route and loadbalancer).
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/area provider/azure
/sig cloud-provider
/priority important-soon

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 1]
```diff
@@ -150,6 +151,8 @@ func getPublicIPDomainNameLabel(service *v1.Service) (string, bool) {
 
 // EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
 func (az *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
+	mc := metrics.NewMetricContext("operation", "ensure_loadbalancer", az.ResourceGroup, az.SubscriptionID, "")
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:154]: let's follow the patterns for other metrics, e.g. change this to `metrics.NewMetricContext("services", "create_or_update", ...)`. And we need to add service name in the metrics

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 2]
```diff
@@ -191,6 +194,7 @@ func (az *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, ser
 		return nil, err
 	}
 
+	mc.ObserveOperation()
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:197]: better to use a defer to catch the earlier returns?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 3]
```diff
@@ -211,6 +215,8 @@ func (az *Cloud) UpdateLoadBalancer(ctx context.Context, clusterName string, ser
 // have multiple underlying components, meaning a Get could say that the LB
 // doesn't exist even if some part of it is still laying around.
 func (az *Cloud) EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error {
+	mc := metrics.NewMetricContext("operation", "ensure_loadbalancer_deleted", az.ResourceGroup, az.SubscriptionID, "")
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:218]: same here

File: staging/src/k8s.io/legacy-cloud-providers/azure/metrics/azure_metrics.go [hunk 4]
```diff
@@ -26,24 +26,38 @@ import (
...
 
-type apiCallMetrics struct {
-	latency          *metrics.HistogramVec
-	errors           *metrics.CounterVec
-	rateLimitedCount *metrics.CounterVec
-	throttledCount   *metrics.CounterVec
-}
+const (
+	azureAPIMetricsNamespace       = "cloudprovider_azure_api_request"
+	azureOperationMetricsNamespace = "cloudprovider_azure_op"
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/metrics/azure_metrics.go:31]: just use `cloudprovider_azure`?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 5]
```diff
@@ -156,6 +157,15 @@ func (az *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, ser
 	serviceName := getServiceName(service)
 	klog.V(5).Infof("ensureloadbalancer(%s): START clusterName=%q", serviceName, clusterName)
 
+	mc := metrics.NewMetricContext("services", "ensure_loadbalancer", az.ResourceGroup, az.SubscriptionID, serviceName)
+	operationSucceeded := false
+	defer func() {
+		mc.ObserveOperation()
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:163]: Could we also add `operationSucceeded` in `ObserveOperation()`, e.g. we could add another function for this purpose, `ObserveOperationWithResult(operationSucceeded)`

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by neolit123** [comment]
please see this PR merge policy change.
https://groups.google.com/g/kubernetes-dev/c/YXGBa6pxLzo

the milestone should be applied later in the cycle.
for Q you can ask in #sig-release.

thanks

/milestone clear

**Comment by nilo19** [comment]
@feiskyer @craiglpeters need lgtm and milestone v1.20 to merge.

**Review by feiskyer — COMMENTED** [review]
/lgtm
/milestone v1.20

**Pull request merged**